src/Workspaces/Core/Portable/Workspace/IsolatedAnalyzerReference.cs

```diff
@@ -0,0 +1,114 @@
+
+    /// <inheritdoc cref="s_analyzerToPinnedReferenceSet"/>
+    private static readonly ConditionalWeakTable<ISourceGenerator, IsolatedAnalyzerReferenceSet> s_generatorToPinnedReferenceSet = [];
+
+    /// <summary>
+    /// We keep a strong reference here.  As long as this <see cref="IsolatedAnalyzerReference"/> is passed out and held
+    /// onto (say by a Project instance), it should keep the IsolatedAssemblyReferenceSet (and its ALC) alive.
+    /// </summary>
+    private readonly IsolatedAnalyzerReferenceSet _isolatedAnalyzerReferenceSet = isolatedAnalyzerReferenceSet;
+
+    /// <summary>
+    /// The actual real <see cref="AnalyzerReference"/> we defer our operations to.
+    /// </summary>
+    public readonly AnalyzerFileReference UnderlyingAnalyzerReference = underlyingAnalyzerReference;
+
+    public override string Display => UnderlyingAnalyzerReference.Display;
+    public override string? FullPath => UnderlyingAnalyzerReference.FullPath;
+    public override object Id => UnderlyingAnalyzerReference.Id;
+
+    public override ImmutableArray<DiagnosticAnalyzer> GetAnalyzers(string language)
+        => PinAnalyzers(static (reference, language) => reference.GetAnalyzers(language), language);
+
+    public override ImmutableArray<DiagnosticAnalyzer> GetAnalyzersForAllLanguages()
+        => PinAnalyzers(static (reference, _) => reference.GetAnalyzersForAllLanguages(), default(VoidResult));
+
+    [Obsolete]
+    public override ImmutableArray<ISourceGenerator> GetGenerators()
+        => PinGenerators(static (reference, _) => reference.GetGenerators(), default(VoidResult));
+
+    public override ImmutableArray<ISourceGenerator> GetGenerators(string language)
+        => PinGenerators(static (reference, language) => reference.GetGenerators(language), language);
+
+    public override ImmutableArray<ISourceGenerator> GetGeneratorsForAllLanguages()
+        => PinGenerators(static (reference, _) => reference.GetGeneratorsForAllLanguages(), default(VoidResult));
+
+    private ImmutableArray<DiagnosticAnalyzer> PinAnalyzers<TArg>(Func<AnalyzerReference, TArg, ImmutableArray<DiagnosticAnalyzer>> getItems, TArg arg)
+        => PinItems(s_analyzerToPinnedReferenceSet, getItems, arg);
+
+    private ImmutableArray<ISourceGenerator> PinGenerators<TArg>(Func<AnalyzerReference, TArg, ImmutableArray<ISourceGenerator>> getItems, TArg arg)
+        => PinItems(s_generatorToPinnedReferenceSet, getItems, arg);
+
+    private ImmutableArray<TItem> PinItems<TItem, TArg>(
+        ConditionalWeakTable<TItem, IsolatedAnalyzerReferenceSet> table,
+        Func<AnalyzerReference, TArg, ImmutableArray<TItem>> getItems,
+        TArg arg)
+        where TItem : class
+    {
+        // Note: we want to keep ourselves alive during this call so that neither we nor our reference set get GC'ed
+        // while we're computing the items.
+        GC.KeepAlive(this);
```

Comment:
@jasonmalinowski i don't think this is necessary.  but if you do, then i'd like to discuss it.

Reply:
that works for me.